src/scroll-timeline-base.js

```diff
@@ -114,9 +114,43 @@ export function calculateMaxScrollOffset(source, orientation) {
     return source.scrollWidth - source.clientWidth;
 }
 
+/**
+ * Calculates the maximum pixel value to use
+ * @param cssValue {CSSUnitValue}
+ * @param source {DOMElement}
+ * @param orientation {String}
+ * @returns {number}
+ */
+function calculateMaxValue(cssValue, source, orientation) {
+  if (cssValue instanceof CSSUnitValue) {
+    switch (cssValue.unit) {
+      case 'vw':
+        return Math.max(document.documentElement.clientWidth || 0, window.innerWidth || 0);
+      case 'vh':
+        return Math.max(document.documentElement.clientHeight || 0, window.innerHeight || 0);
```

Comment:
I think this will return the size of the dynamic viewport on iOS Safari, and not the large viewport?
I wish there was an js api for reading these viewport relative units.

Reply:
The `document.documentElement.client*` properties return the size of the ICB[^1]. In practice – and soon [in the spec too](https://github.com/w3c/csswg-drafts/issues/6453#issuecomment-1789821736) – this is the same value as reported by the the `sv*` units, i.e. when the dynamic browser UI is expanded. 

The Layout Viewport _typically_ equals the values reported by `window.innerWidth`/`window.innerHeight`. When the dynamic browser UI gets retracted, these values grow as more space becomes available[^2]. 

So by theory you could just use use `window.innerWidth`/`window.innerHeight`. In practice there is a bug in iOS Safari where these values shrink when having pinch-zoomed in[^3]. WebKit on mobile is the only engine that does this.

That should explain why there is a fallback to the `document.documentElement.client*` values. It’s not 100% correct when having pinch-zoomed in, but at least it‘s not entirely wrong.


[^1]: Explainer: https://github.com/web-platform-tests/interop-2022-viewport/blob/main/explainers/icb.md
[^2]: Demo Page: https://interop-2022-viewport.netlify.app/individual/layout-viewport/
[^3]: Bug: https://bugs.webkit.org/show_bug.cgi?id=174362

Reply:
> The `document.documentElement.client*` properties return the size of the ICB[1](#user-content-fn-1-4a745e109dc1cbe47cfa6d69e68965fe). In practice – and soon [in the spec too](https://github.com/w3c/csswg-drafts/issues/6453#issuecomment-1789821736) – this is the same value as reported by the the `sv*` units, i.e. when the dynamic browser UI is expanded.

Nice. I didn't catch that resolution. 👍 
I appreciate the write-up 🙏

We have a way to resolve the small viewport size by evaluating `document.documentElement.clientHeight`. Evaluating `window.innerHeight` will give us the dynamic viewport size. (Apart from the bugs) 

However, I don't see that we have a way to resolve the large viewport size though? Without injecting something into the dom to measure it?

```js
Math.max(document.documentElement.clientHeight || 0, window.innerHeight || 0);
```
The expression above will resolve to the size of the dynamic viewport? (And at times be smaller than the large viewport size)

Reply:
> However, I don't see that we have a way to resolve the large viewport size though? Without injecting something into the dom to measure it?

Correct.

> The expression above will resolve to the size of the dynamic viewport? (And at times be smaller than the large viewport size)

Also correct _(except for WebKit on Mobile when pinch-zoomed in due to that bug they have)_, so the code needs to be updated (for mobile).

Quickly worked out this that should get all values. Doesn’t include the fix for WebKit on Mobile though: https://codepen.io/bramus/pen/dyaJOza?editors=0010

Reply:
Note with that demo/code linked above: only seems to give correct results in MobileSafari (on iOS 17.0.3) itself. Chrome/Edge/Firefox give wrong results, most likely because they are a limited WebKit under the hood.

Reply:
> Note with that demo/code linked above: only seems to give correct results in MobileSafari (on iOS 17.0.3) itself. Chrome/Edge/Firefox give wrong results, most likely because they are a limited WebKit under the hood.

Yeah.. Viewport-percentage lengths are still unreliable in many WebViews and In App Browsers 😞.

Reply:
Just came to the realization that Viewport Units in a WebKit WebView on iOS 17 are entirely broken. This wasn't the case on iOS 15.

Test Page: https://interop-2022-viewport.netlify.app/combined/viewport-units/